apps/sync/js/adapters/bookmarks.js

```diff
@@ -0,0 +1,334 @@
+    return new Promise(resolve => {
+      asyncStorage.setItem(BOOKMARKS_COLLECTION_MTIME, mtime, resolve);
+    });
+  }
+
+  function getSyncedCollectionMtime() {
+    return new Promise(resolve => {
+      asyncStorage.getItem(BOOKMARKS_COLLECTION_MTIME, resolve);
+    });
+  }
+
+  function setDataStoreId(synctoId, dataStoreId) {
+    return new Promise(resolve => {
+      asyncStorage.setItem(
+        BOOKMARKS_SYNCTOID_PREFIX + synctoId, dataStoreId, resolve);
+    });
+  }
+
+  function getDataStoreId(synctoId) {
+    return new Promise(resolve => {
+      asyncStorage.getItem(BOOKMARKS_SYNCTOID_PREFIX + synctoId, resolve);
+    });
+  }
+
+  function mergeRecordsToDataStore(localRecord, remoteRecord) {
+    if (!localRecord || !remoteRecord ||
+        localRecord.id !== remoteRecord.id ||
+        (remoteRecord.type === 'url' && localRecord.url !== remoteRecord.url)) {
+      // Local and remote records have different url(id).
+      console.error('Inconsistent records', localRecord, remoteRecord);
+      throw new Error('Inconsistent records');
+    }
+    if (!localRecord.fxsyncId && typeof remoteRecord.fxsyncId === 'string') {
+      /* When a localRecord is existed without fxsyncId, assign fxsyncId to it
+         from a remoteRecord. This case always happens at first synchronization
+         or merging two records with the same URL. */
+      localRecord.fxsyncId = remoteRecord.fxsyncId;
+    } else if (localRecord.fxsyncId !== remoteRecord.fxsyncId) {
+      // Two records have different fxsyncId but have the same url(id).
+      console.log('Records should have the same Firefox Sync ID',
+        localRecord, remoteRecord);
+      throw new Error('Records should have the same Firefox Sync ID',
+        localRecord, remoteRecord);
+    }
+
+    localRecord.name = remoteRecord.name;
+    localRecord.fxsyncPayload = remoteRecord.fxsyncPayload;
+
+    return localRecord;
+  }
+
+  function addBookmark(remoteRecord) {
+    // 1. Get bookmark by url(id of DataStore)
+    // 2.A. If the bookmark already exists locally,
+    //     we merge it with the remote one.
+    // 2.B Add a new record with RevisionId.
+    // 3. Add the DataStore record ID into LocalID <-> RemoteID matching table.
+
+    var id = remoteRecord.id;
+    var revisionId;
+    return _ensureStore().then(store => {
+      revisionId = store.revisionId;
+      return store.get(id);
+    }).then(localRecord => {
+      if (localRecord) {
+        var newBookmark = mergeRecordsToDataStore(localRecord, remoteRecord);
+        return store.put(newBookmark, id, revisionId);
+      }
+      return store.add(remoteRecord, id, revisionId).then(() => {
+        return setDataStoreId(remoteRecord.fxsyncId, id);
+      });
+    }).catch(e => {
+      console.error(e);
+    });
+  }
+
+  function updateBookmarks(records) {
+    return new Promise(resolve => {
+      records.reduce((reduced, current) => {
+        return reduced.then(() => {
```

Comment:
I think you should set initialValue to `Promise.resolve()` here? Otherwise I think it will fail of `records.length <= 1`? See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/Reduce#Description